src/core/fetchers/segment/promise_prioritizer.ts

```diff
@@ -0,0 +1,321 @@
+import log from "../../../log";
+import arrayFindIndex from "../../../utils/array_find_index";
+import TaskCanceller from "../../../utils/task_canceller";
+
+export default class PromisePrioritizer<T> {
+  private _minPendingPriority: number | null;
+  private _waitingQueue: Array<IPrioritizerTask<T>>;
+  private _pendingTasks: Array<IPrioritizerTask<T>>;
+  private _prioritySteps: IPrioritizerPrioritySteps;
+
+  constructor({ prioritySteps }: IPrioritizerOptions) {
+    this._minPendingPriority = null;
+    this._waitingQueue = [];
+    this._pendingTasks = [];
+    this._prioritySteps = prioritySteps;
+
+    if (this._prioritySteps.high >= this._prioritySteps.low) {
+      throw new Error(
+        "FP Error: the max high level priority should be given a lower" +
+          "priority number than the min low priority."
+      );
+    }
+  }
+
+  public create(
+    promise: () => Promise<T>,
+    priority: number,
+    canceller: TaskCanceller
```

Comment:
I don't understand the point of giving `create` a `TaskCanceller`.
A `CancellationSignal` I would get it (it would be to listen to cancel order).